src/mml_utils.py

```diff
@@ -0,0 +1,129 @@
+from bayeslite import bql_quote_name
+from bdbcontrib.population_method import population_method
+from enum import Enum
+from string import Template
+
+import json
+import jsonschema
+import pkgutil
+
+MML_SCHEMA = json.loads(
+    pkgutil.get_data('bdbcontrib', 'mml.schema.json'))
+
+
+class StatType(Enum):
+    CATEGORICAL = 1
+    NUMERICAL = 2
+    IGNORE = 3
+
+
+@population_method(population_to_bdb=0, population_name=1)
+def guess_types(bdb, table):
+    """Guesses stattypes of a given table.
+
+    Returns a dictionary from column names to the guessed stattype.
+    """
+    types = {}
+    for col in _column_names(bdb, table):
+        cursor = bdb.sql_execute(
+            'SELECT DISTINCT %s FROM %s'
+            % (bql_quote_name(col), bql_quote_name(table)))
+        # TODO(asilvers): We don't necessarily need to read all of these in,
+        # and for some datasets this may be prohibitive. But it's fine for now.
+        vals = {row[0] for row in cursor if row[0] not in [None, '']}
+        types[col] = _type_given_vals(vals)
+    return types
+
+
+def _column_names(bdb, table):
+    """Returns the column names of a table"""
+    with bdb.savepoint():
+        cursor = bdb.sql_execute('PRAGMA table_info(%s)'
+                                 % (bql_quote_name(table),))
+        return list(row[1] for row in cursor)
+
+
+def _type_given_vals(vals):
+    """Returns a guess as to the stattype of a column given its values.
+
+    These heuristics are utterly unprincipled and simply seem to work well
+    on a minimal set of test datasets. They should be treated fluidly and
+    improved as failures crop up.
+    """
```

Comment:
Yes...  I'm not addressing that: I agree the automatic heuristics don't work very well and pretty much everyone will want to tweak them manually.  Rather, I'm addressing the *additional* heuristics added by this change, and suggesting that instead of adding new heuristics, it could just use the same ones that `GUESS(*)` uses, by calling `bayesdb_guess_stattypes` from `bayeslite.guess`.